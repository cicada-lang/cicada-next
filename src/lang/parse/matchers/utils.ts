import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
