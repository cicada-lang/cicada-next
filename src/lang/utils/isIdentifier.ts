const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
