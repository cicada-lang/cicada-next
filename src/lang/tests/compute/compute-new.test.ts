import { expect, test } from "vitest"
import { runCode } from "../utils"

test("compute New", async () => {
  const output = await runCode(`

class ABC {
  a: String
  b: String
  c: String
}

compute new ABC {
  a: "a",
  b: "b",
  c: "c",
}

`)

  expect(output).toMatchInlineSnapshot(
    '"{ a: \\"a\\", b: \\"b\\", c: \\"c\\" }: class { a: String, b: String, c: String }"',
  )
})

test("compute New -- fulfilled", async () => {
  const output = await runCode(`

class AB {
  a: String
  b: String = a
}

compute new AB {
  a: "Hello, World!",
}

`)

  expect(output).toMatchInlineSnapshot(
    '"{ a: \\"Hello, World!\\", b: \\"Hello, World!\\" }: class { a: String, b: String = a }"',
  )
})

test("compute New -- dependent", async () => {
  const output = await runCode(`

class The {
  T: Type
  x: T
}

compute new The {
  T: String,
  x: "x",
}

`)

  expect(output).toMatchInlineSnapshot('"{ T: String, x: \\"x\\" }: class { T: Type, x: T }"')
})

test("compute New -- extra properties", async () => {
  const output = await runCode(`

class ABC {
  a: String
  b: String
  c: String
}

compute new ABC {
  a: "a",
  b: "b",
  c: "c",
  d: "d",
}

`)

  expect(output).toMatchInlineSnapshot(
    '"{ d: \\"d\\", a: \\"a\\", b: \\"b\\", c: \\"c\\" }: class { d: String = \\"d\\", a: String, b: String, c: String }"',
  )
})
