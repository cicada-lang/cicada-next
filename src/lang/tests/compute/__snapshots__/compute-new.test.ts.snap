// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compute New -- dependent 1`] = `
"{
  T: String,
  x: "x"
}: class The {
  T: Type = String
  x: String = "x"
}"
`;

exports[`compute New -- extra properties 1`] = `
"{
  a: "a",
  b: "b",
  c: "c",
  d: "d"
}: class ABC {
  a: String = "a"
  b: String = "b"
  c: String = "c"
  d: String = "d"
}"
`;

exports[`compute New -- fulfilled 1`] = `
"{
  a: "Hello, World!",
  b: "Hello, World!"
}: class AB {
  a: String = "Hello, World!"
  b: String = "Hello, World!"
}"
`;

exports[`compute New 1`] = `
"{
  a: "a",
  b: "b",
  c: "c"
}: class ABC {
  a: String = "a"
  b: String = "b"
  c: String = "c"
}"
`;
